### ISSUE TITLE
Find by inode number

### ISSUE THREAD

**Issue by sucrecacao** [sharkdp/fd#880]
Hi!

Thanks for this amazing tool :) I use it everyday

legacy `find` has an option to search by inode number ( `find -inum XXXXX`)  but it is terribly slow.

Would it be possible to implement this feature in `fd`, and if so, would it be significantly faster than `find` ?

**Comment by tavianator** [comment]
Can you describe your use case for this?  Is it something like #300/#751?

I don't think we're likely to implement something like `-inum`, and `--samefile` has been rejected before.

**Comment by sucrecacao** [comment]
>  Is it something like #300/#751?

No

> Can you describe your use case for this?

Searching by inode can be extremely usefull in various usecase, and it is a general purpose rather than specific usecase tool. But where I find it most usefull, is to keep track of files in media library ( image, music, video, torrent). A user can freely move files around and  rename them; the software will still be able to find the file, even if its path or name has changed.

**Comment by sucrecacao** [comment]
Would it be easy to print the inode instead of the filepath in `fd` ? 

Then it can be used with `grep` to quickly find the inode

**Comment by tavianator** [comment]
> Searching by inode can be extremely usefull in various usecase, and it is a general purpose rather than specific usecase tool.

Sure, but features are only added to `fd` if they have compelling usecases.  I personally have never used `find -inum` except to test that it works.  So I was curious why others find it useful.

> But where I find it most usefull, is to keep track of files in media library ( image, music, video, torrent). A user can freely move files around and rename them; the software will still be able to find the file, even if its path or name has changed.

I guess my question is, how do you even know the inode number of the file you're looking for?  Do you memorize them?  Or do you have a database that stores them?

> Would it be easy to print the inode instead of the filepath in fd ?  Then it can be used with grep to quickly find the inode

It would be possible.  I don't think it's a feature we're likely to implement though.  You could try `fd -X ls -1di` or similar to have `ls` do it for you, but I'm not sure how performant that is.

**Comment by sucrecacao** [comment]
>  I personally have never used find -inum except to test that it works. So I was curious why others find it useful.

yes, you would not definitly never use as you use `fd` for working workflow. You would most probably use this feature inside a software or a script only. 

> I guess my question is, how do you even know the inode number of the file you're looking for? Do you memorize them? Or do you have a database that stores them?

Yes, you track the inode of set of files (e.g. when adding a torrent, when import photo/music to you media library ). You store there inode and path. If there are not located at the usual path, you start inode search. 

> , but I'm not sure how performant that is.

I don't know how `fd` achieve to be so much faster than `find`, but I guess its not because it has algorithm with best time complexity, `fd` just execute faster. that would be great to have the same benefice for inode search, but I understand it is not top priority.

**Comment by Sherlock-Holo** [comment]
@tavianator 
if we got a cgroupv2 id, to find out the cgroup path, the good way is `find -inum` because the cgroupv2 id is the inode https://github.com/iovisor/bpftrace/issues/1500#issuecomment-690750082

I think that may be a common use case

**Comment by cgzones** [comment]
Another use case is to find the file related to an audit or SELinux event only referenced by their inode number.

**Comment by tavianator** [comment]
I don't think this feature belongs in `fd`.  If you need something faster than `find`, [`bfs`](https://github.com/tavianator/bfs) supports `-inum`.  Some filesystems support mapping inode numbers to paths directly, which should be even faster, e.g.

```console
$ sudo btrfs inspect-internal inode-resolve 9253882 /
//home/tavianator/code/sharkdp/fd/target/release/fd
```

https://unix.stackexchange.com/q/35292/56202 has more examples for other filesystems.

**Comment by webstrand** [comment]
I was looking for this, my use case is finding files that `fsck` reported as damaged, since it only gives me their inode.
